File: dart/data_structures/heap/min_heap.dart
// min_heap.dart

class MinHeap {
  List<int> _heap;

  MinHeap() : _heap = [];

  void insert(int value) {
    _heap.add(value);
    _heapifyUp();
  }

  void _heapifyUp() {
    int index = _heap.length - 1;

    while (index > 0) {
      int parentIndex = (index - 1) ~/ 2;

      if (_heap[index] < _heap[parentIndex]) {
        _swap(index, parentIndex);
        index = parentIndex;
      } else {
        break;
      }
    }
  }

  int extractMin() {
    if (_heap.isEmpty) {
      throw StateError("Heap is empty");
    }

    int minValue = _heap[0];

    if (_heap.length == 1) {
      _heap.removeAt(0);
    } else {
      _heap[0] = _heap.removeLast();
      _heapifyDown();
    }

    return minValue;
  }

  void _heapifyDown() {
    int index = 0;

    while (true) {
      int leftChildIndex = 2 * index + 1;
      int rightChildIndex = 2 * index + 2;
      int smallestChildIndex = index;

      if (leftChildIndex < _heap.length &&
          _heap[leftChildIndex] < _heap[smallestChildIndex]) {
        smallestChildIndex = leftChildIndex;
      }

      if (rightChildIndex < _heap.length &&
          _heap[rightChildIndex] < _heap[smallestChildIndex]) {
        smallestChildIndex = rightChildIndex;
      }

      if (smallestChildIndex != index) {
        _swap(index, smallestChildIndex);
        index = smallestChildIndex;
      } else {
        break;
      }
    }
  }

  void _swap(int i, int j) {
    int temp = _heap[i];
    _heap[i] = _heap[j];
    _heap[j] = temp;
  }

  void display() {
    print("Min Heap: $_heap");
  }
}

void main() {
  MinHeap minHeap = MinHeap();

  minHeap.insert(3);
  minHeap.insert(1);
  minHeap.insert(4);
  minHeap.insert(2);

  print("Original Min Heap:");
  minHeap.display();

  print("\nExtract Min Value: ${minHeap.extractMin()}");

  print("Min Heap after Extraction:");
  minHeap.display();
}

File: go/data_structures/heap/min_heap.go
// min_heap.go

package main

import "fmt"

type MinHeap struct {
    heap []int
}

func NewMinHeap() *MinHeap {
    return &MinHeap{}
}

func (h *MinHeap) Insert(value int) {
    h.heap = append(h.heap, value)
    h.heapifyUp(len(h.heap) - 1)
}

func (h *MinHeap) ExtractMin() (int, error) {
    if len(h.heap) == 0 {
        return 0, fmt.Errorf("heap is empty")
    }

    min := h.heap[0]
    lastIdx := len(h.heap) - 1

    h.heap[0], h.heap[lastIdx] = h.heap[lastIdx], h.heap[0]
    h.heap = h.heap[:lastIdx]

    h.heapifyDown(0)

    return min, nil
}

func (h *MinHeap) heapifyUp(index int) {
    parent := (index - 1) / 2

    for index > 0 && h.heap[index] < h.heap[parent] {
        h.heap[index], h.heap[parent] = h.heap[parent], h.heap[index]
        index = parent
        parent = (index - 1) / 2
    }
}

func (h *MinHeap) heapifyDown(index int) {
    for {
        leftChild := 2*index + 1
        rightChild := 2*index + 2
        smallest := index

        if leftChild < len(h.heap) && h.heap[leftChild] < h.heap[smallest] {
            smallest = leftChild
        }

        if rightChild < len(h.heap) && h.heap[rightChild] < h.heap[smallest] {
            smallest = rightChild
        }

        if smallest != index {
            h.heap[index], h.heap[smallest] = h.heap[smallest], h.heap[index]
            index = smallest
        } else {
            break
        }
    }
}

func main() {
    heap := NewMinHeap()

    heap.Insert(4)
    heap.Insert(10)
    heap.Insert(3)
    heap.Insert(5)
    heap.Insert(1)

    fmt.Println("Min Heap:", heap.heap)

    min, err := heap.ExtractMin()
    if err == nil {
        fmt.Printf("Extracted Min Element: %d\n", min)
        fmt.Println("Updated Min Heap:", heap.heap)
    } else {
        fmt.Println(err)
    }
}
